plugins/backup/veeam/src/main/java/org/apache/cloudstack/backup/VeeamBackupProvider.java

```diff
@@ -0,0 +1,319 @@
+    }
+
+    @Override
+    public boolean isBackupOffering(final Long zoneId, final String uuid) {
+        List<BackupOffering> policies = listBackupOfferings(zoneId);
+        if (CollectionUtils.isEmpty(policies)) {
+            return false;
+        }
+        for (final BackupOffering policy : policies) {
+            if (policy.getExternalId().equals(uuid)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private VmwareDatacenter findVmwareDatacenterForVM(final VirtualMachine vm) {
+        if (vm == null || vm.getHypervisorType() != Hypervisor.HypervisorType.VMware) {
+            throw new CloudRuntimeException("The Veeam backup provider is only applicable for VMware VMs");
+        }
+        final VmwareDatacenterZoneMap zoneMap = vmwareDatacenterZoneMapDao.findByZoneId(vm.getDataCenterId());
+        if (zoneMap == null) {
+            throw new CloudRuntimeException("Failed to find a mapped VMware datacenter for zone id:" + vm.getDataCenterId());
+        }
+        final VmwareDatacenter vmwareDatacenter = vmwareDatacenterDao.findById(zoneMap.getVmwareDcId());
+        if (vmwareDatacenter == null) {
+            throw new CloudRuntimeException("Failed to find a valid VMware datacenter mapped for zone id:" + vm.getDataCenterId());
+        }
+        return vmwareDatacenter;
+    }
+
+    public boolean addVMToBackupOffering(final BackupOffering policy, final VirtualMachine vm) {
+        final VmwareDatacenter vmwareDatacenter = findVmwareDatacenterForVM(vm);
+        return getClient(vm.getDataCenterId()).removeVMFromVeeamJob(policy.getExternalId(), vm.getInstanceName(), vmwareDatacenter.getVcenterHost());
+    }
+
+    private String getGuestBackupName(final String instanceName, final String uuid) {
+        return String.format("%s%s%s", instanceName, BACKUP_IDENTIFIER, uuid);
+    }
+
+    @Override
+    public boolean assignVMToBackupOffering(final VirtualMachine vm, final BackupOffering backupOffering) {
+        final VeeamClient client = getClient(vm.getDataCenterId());
+        final Job parentJob = client.listJob(backupOffering.getExternalId());
+        final String clonedJobName = getGuestBackupName(vm.getInstanceName(), vm.getUuid());
+        if (client.cloneVeeamJob(parentJob, clonedJobName)) {
+            for (BackupOffering job : client.listJobs()) {
+                if (job.getName().equals(clonedJobName)) {
+                    final Job clonedJob = client.listJob(job.getExternalId());
+                    if (clonedJob.getScheduleConfigured() && !clonedJob.getScheduleEnabled()) {
+                        client.toggleJobSchedule(clonedJob.getId());
+                    }
+                    final VmwareDatacenter vmwareDC = findVmwareDatacenterForVM(vm);
+                    if (client.addVMToVeeamJob(job.getExternalId(), vm.getInstanceName(), vmwareDC.getVcenterHost())) {
+                        ((VMInstanceVO) vm).setBackupExternalId(job.getExternalId());
```

Comment:
Can we move this BackupManager and set the external ID if the operation was successful?

Reply:
The problem here is that the external ID is something specific to veeam provider; the job object has it. We can refactor this in future.